pkg/kubectl/cmd/taint.go

```diff
@@ -113,42 +120,56 @@ func NewCmdTaint(f cmdutil.Factory, out io.Writer) *cobra.Command {
 
 // reorganizeTaints returns the updated set of taints, taking into account old taints that were not updated,
 // old taints that were updated, old taints that were deleted, and new taints.
-func reorganizeTaints(obj runtime.Object, overwrite bool, taintsToAdd []v1.Taint, taintsToRemove []v1.Taint) ([]v1.Taint, error) {
-	node, ok := obj.(*v1.Node)
-	if !ok {
-		return nil, fmt.Errorf("unexpected type %T, expected Node", obj)
-	}
-
+func reorganizeTaints(node *v1.Node, overwrite bool, taintsToAdd []v1.Taint, taintsToRemove []v1.Taint) (string, []v1.Taint, error) {
 	newTaints := append([]v1.Taint{}, taintsToAdd...)
-
 	oldTaints := node.Spec.Taints
 	// add taints that already existing but not updated to newTaints
-	for _, oldTaint := range oldTaints {
-		existsInNew := false
-		for _, taint := range newTaints {
-			if taint.MatchTaint(&oldTaint) {
-				existsInNew = true
-				break
-			}
-		}
-		if !existsInNew {
-			newTaints = append(newTaints, oldTaint)
-		}
+	added := addTaints(oldTaints, &newTaints)
+	allErrs, deleted := deleteTaints(taintsToRemove, &newTaints)
+	if added && deleted {
+		return MODIFIED, newTaints, utilerrors.NewAggregate(allErrs)
+	} else if added {
+		return TAINTED, newTaints, utilerrors.NewAggregate(allErrs)
+	} else if overwrite {
+		return OVERWRITTEN, newTaints, utilerrors.NewAggregate(allErrs)
 	}
+	return UNTAINTED, newTaints, utilerrors.NewAggregate(allErrs)
+}
 
+// deleteTaints deletes the given taints from the node's taintlist.
+func deleteTaints(taintsToRemove []v1.Taint, newTaints *[]v1.Taint) ([]error, bool) {
 	allErrs := []error{}
+	taintsBefore := len(*newTaints)
 	for _, taintToRemove := range taintsToRemove {
 		removed := false
 		if len(taintToRemove.Effect) > 0 {
-			newTaints, removed = v1.DeleteTaint(newTaints, &taintToRemove)
+			*newTaints, removed = v1.DeleteTaint(*newTaints, &taintToRemove)
 		} else {
-			newTaints, removed = v1.DeleteTaintsByKey(newTaints, taintToRemove.Key)
+			*newTaints, removed = v1.DeleteTaintsByKey(*newTaints, taintToRemove.Key)
 		}
 		if !removed {
 			allErrs = append(allErrs, fmt.Errorf("taint %q not found", taintToRemove.ToString()))
 		}
 	}
-	return newTaints, utilerrors.NewAggregate(allErrs)
+	return allErrs, taintsBefore != len(*newTaints)
```

Comment:
So, here the errors are different and taints are different, we are returning errors as well as checking if the original length of taints list is equal to old length of taints to check if some of them got deleted.